### DISCUSSION TITLE
sympy. physics.mechanics //auxiliary speeds

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#25020] [Q&A]
If I want to calculate the reaction forces acting on some point, I give this point 'virtual' speeds, say P1.set_vel(N, aux_x * N.x + aux_y * N.y + auxz * N.z), and setting up the force list I put
(P1, f_x * N.x + f_y * N.y + f_z * N.z)

'Normally' these symbols aux_i, f_i do not appear in the force term - but sometimes they do appear.
As they do not work on the system, I simply set them to zero.

As I have no idea about the inner workings of sympy mechanics, my question, just curiosity, is:
Why do they appear sometimes?

It seems to me, that if a point has real and virtual speeds, they do appear, but this is just observation, and does not seem yo hold always.

Any explanation would be greatly appreciated!

**Comment by moorepants** [comment]
You may be hitting a bug. The auxiliary speeds and maybe even the forces aren't eliminated from the equations of motion by KanesMethod, but the should be. You will have to manually set them to zero. Might that be what you are seeing?

**Reply by Peter230655** [reply]
Yes, this is what I am seeing. So no harm, I just set them to zero?

**Reply by moorepants** [reply]
Yes, but we need to fix this behavior in SymPy. We should eliminate all of those automatically. I've been working on this new bicycle model and you can see I have to do the same thing: https://github.com/mechmotum/bicycle-kickplate-model/blob/master/nonlin_sym.py#L553

**Reply by Peter230655** [reply]
Further to my point above, I ran into something, new for me anyway, right now:

If I want to calculate the reaction forces, I use
_reaction_force = KM.auxiliary_eqs_
which I then solve numerically for the rection forces.

Right now, in my example, I found that the virtual speeds, aux_i appear as dynamic symbols in the term reaction_forces.

I have not seen this happening before. 
Can I simply set them to zero here, too?

Thanks for any help!

**Reply by moorepants** [reply]
Yes you can.

**Reply by Peter230655** [reply]
I see it. Thanks!
As long as setting aux and also reaction forces = zero is fine, annoying but ultimately harmless.

Maybe after my retirement I will be able to learn to

**Comment by moorepants** [comment]
The resulting equations should look like this:

$M(q) \dot{u} + N r = F(u, q)$
$A(u, q) r + B(q) \dot{u} = G(u, q)$

No aux speeds or their derivatives should be present and $r$ are the constraint forces.

**Reply by Peter230655** [reply]
Thanks!
Are 'my' two issues related?
I have seen the virtual speedsin the force vector before, but in the reaction_force it was the first time.

**Reply by moorepants** [reply]
It is the same issue. KanesMethod does not replace the auxiliary speeds and their derivatives with zero and it should.

**Reply by moorepants** [reply]
Here is an open issue about it https://github.com/sympy/sympy/issues/13635

**Comment by Peter230655** [comment]
In my proposed change to kane.py for nonlinear motion constrraints I eliminate the virtual speeds in the force vector and in the auxiliary_eqs - but only if nonlinear speed constraints are present. Two reasons for this:
- I wanted to change the original as little as possible.
- It _seems_ to happen more often when nonlinear speed constraints are used.
- Removing terms can be time consuming if the expression where the terms are to be removed is large.
 
I do not know how to eliminate the reaction forces because I do not know how to identify them within kane.py. So, I would need a new keyword just for this purpose, which does not seem to make sense.
**Can this be closed?**

**Comment by Peter230655** [comment]
I see no way to eliminate the noncontributing forces in the force vector, unless we add asn optional keyword non_contributing_forces=[names of the forces]
Of course it is a minor inconveninece to eliminate the from the force vector later, so maybe all that's needed is a remark in the documentation.
